File: sdks/java/src/main/java/do_/mongo/Indexes.java
package do_.mongo;

import java.util.Arrays;
import java.util.List;

/**
 * Utility class for creating MongoDB index specifications.
 * <p>
 * This class provides static factory methods for building index key documents.
 * </p>
 *
 * <pre>{@code
 * import static do_.mongo.Indexes.*;
 *
 * // Single field ascending
 * Bson index = ascending("email");
 *
 * // Compound index
 * Bson index = compoundIndex(ascending("lastName"), descending("age"));
 *
 * // Text index
 * Bson index = text("content");
 *
 * // Geospatial index
 * Bson index = geo2dsphere("location");
 *
 * // Create with options
 * collection.createIndex(ascending("email"),
 *     new Document("unique", true).append("sparse", true));
 * }</pre>
 */
public final class Indexes {

    private Indexes() {
        // Utility class
    }

    /**
     * Creates an ascending index specification for a single field.
     *
     * @param fieldName the field name
     * @return the index specification
     */
    public static Bson ascending(String fieldName) {
        return () -> new Document(fieldName, 1);
    }

    /**
     * Creates an ascending index specification for multiple fields.
     *
     * @param fieldNames the field names
     * @return the index specification
     */
    public static Bson ascending(String... fieldNames) {
        return ascending(Arrays.asList(fieldNames));
    }

    /**
     * Creates an ascending index specification for multiple fields.
     *
     * @param fieldNames the field names
     * @return the index specification
     */
    public static Bson ascending(List<String> fieldNames) {
        return () -> {
            Document index = new Document();
            for (String field : fieldNames) {
                index.append(field, 1);
            }
            return index;
        };
    }

    /**
     * Creates a descending index specification for a single field.
     *
     * @param fieldName the field name
     * @return the index specification
     */
    public static Bson descending(String fieldName) {
        return () -> new Document(fieldName, -1);
    }

    /**
     * Creates a descending index specification for multiple fields.
     *
     * @param fieldNames the field names
     * @return the index specification
     */
    public static Bson descending(String... fieldNames) {
        return descending(Arrays.asList(fieldNames));
    }

    /**
     * Creates a descending index specification for multiple fields.
     *
     * @param fieldNames the field names
     * @return the index specification
     */
    public static Bson descending(List<String> fieldNames) {
        return () -> {
            Document index = new Document();
            for (String field : fieldNames) {
                index.append(field, -1);
            }
            return index;
        };
    }

    /**
     * Creates a compound index from multiple index specifications.
     *
     * @param indexes the index specifications
     * @return the compound index specification
     */
    public static Bson compoundIndex(Bson... indexes) {
        return compoundIndex(Arrays.asList(indexes));
    }

    /**
     * Creates a compound index from multiple index specifications.
     *
     * @param indexes the index specifications
     * @return the compound index specification
     */
    public static Bson compoundIndex(Iterable<Bson> indexes) {
        return () -> {
            Document combined = new Document();
            for (Bson index : indexes) {
                combined.putAll(index.toBsonDocument());
            }
            return combined;
        };
    }

    /**
     * Creates a text index specification.
     *
     * @param fieldName the field name
     * @return the index specification
     */
    public static Bson text(String fieldName) {
        return () -> new Document(fieldName, "text");
    }

    /**
     * Creates a text index specification for multiple fields.
     *
     * @param fieldNames the field names
     * @return the index specification
     */
    public static Bson text(String... fieldNames) {
        return text(Arrays.asList(fieldNames));
    }

    /**
     * Creates a text index specification for multiple fields.
     *
     * @param fieldNames the field names
     * @return the index specification
     */
    public static Bson text(List<String> fieldNames) {
        return () -> {
            Document index = new Document();
            for (String field : fieldNames) {
                index.append(field, "text");
            }
            return index;
        };
    }

    /**
     * Creates a wildcard text index for all string fields.
     *
     * @return the index specification
     */
    public static Bson textAll() {
        return () -> new Document("$**", "text");
    }

    /**
     * Creates a hashed index specification.
     *
     * @param fieldName the field name
     * @return the index specification
     */
    public static Bson hashed(String fieldName) {
        return () -> new Document(fieldName, "hashed");
    }

    /**
     * Creates a 2dsphere geospatial index.
     *
     * @param fieldName the field name
     * @return the index specification
     */
    public static Bson geo2dsphere(String fieldName) {
        return () -> new Document(fieldName, "2dsphere");
    }

    /**
     * Creates a 2dsphere geospatial index for multiple fields.
     *
     * @param fieldNames the field names
     * @return the index specification
     */
    public static Bson geo2dsphere(String... fieldNames) {
        return () -> {
            Document index = new Document();
            for (String field : fieldNames) {
                index.append(field, "2dsphere");
            }
            return index;
        };
    }

    /**
     * Creates a 2d geospatial index (for legacy coordinate pairs).
     *
     * @param fieldName the field name
     * @return the index specification
     */
    public static Bson geo2d(String fieldName) {
        return () -> new Document(fieldName, "2d");
    }

    /**
     * Creates a wildcard index.
     *
     * @return the index specification for all fields
     */
    public static Bson wildcard() {
        return () -> new Document("$**", 1);
    }

    /**
     * Creates a wildcard index for a specific path.
     *
     * @param fieldPath the field path (e.g., "metadata.$**")
     * @return the index specification
     */
    public static Bson wildcard(String fieldPath) {
        String path = fieldPath.endsWith("$**") ? fieldPath : fieldPath + ".$**";
        return () -> new Document(path, 1);
    }

    // ============================================================================
    // Index Options Helpers
    // ============================================================================

    /**
     * Creates index options for a unique index.
     *
     * @return the options document
     */
    public static Document uniqueOptions() {
        return new Document("unique", true);
    }

    /**
     * Creates index options for a sparse index.
     *
     * @return the options document
     */
    public static Document sparseOptions() {
        return new Document("sparse", true);
    }

    /**
     * Creates index options for a unique sparse index.
     *
     * @return the options document
     */
    public static Document uniqueSparseOptions() {
        return new Document("unique", true).append("sparse", true);
    }

    /**
     * Creates index options with a custom name.
     *
     * @param name the index name
     * @return the options document
     */
    public static Document namedOptions(String name) {
        return new Document("name", name);
    }

    /**
     * Creates index options for a TTL index.
     *
     * @param expireAfterSeconds the time in seconds after which documents expire
     * @return the options document
     */
    public static Document ttlOptions(long expireAfterSeconds) {
        return new Document("expireAfterSeconds", expireAfterSeconds);
    }

    /**
     * Creates index options for a partial index.
     *
     * @param filter the filter expression
     * @return the options document
     */
    public static Document partialOptions(Bson filter) {
        return new Document("partialFilterExpression", filter.toBsonDocument());
    }

    /**
     * Creates index options for a background build.
     *
     * @return the options document
     */
    public static Document backgroundOptions() {
        return new Document("background", true);
    }

    /**
     * Combines multiple index options.
     *
     * @param options the options to combine
     * @return the combined options document
     */
    public static Document combineOptions(Document... options) {
        Document combined = new Document();
        for (Document opt : options) {
            combined.putAll(opt);
        }
        return combined;
    }

    /**
     * Creates index options for a text index with weights.
     *
     * @param weights the field weights
     * @return the options document
     */
    public static Document textWeights(Document weights) {
        return new Document("weights", weights);
    }

    /**
     * Creates index options for a text index with language.
     *
     * @param language the default language
     * @return the options document
     */
    public static Document textLanguage(String language) {
        return new Document("default_language", language);
    }

    /**
     * Creates index options for a text index with a language override field.
     *
     * @param field the field that specifies document language
     * @return the options document
     */
    public static Document textLanguageOverride(String field) {
        return new Document("language_override", field);
    }

    /**
     * Creates index options for collation.
     *
     * @param collation the collation document
     * @return the options document
     */
    public static Document collationOptions(Document collation) {
        return new Document("collation", collation);
    }
}

File: sdks/kotlin/src/main/kotlin/do_/mongo/Indexes.kt
package do_.mongo

/**
 * Index key builders for MongoDB.
 *
 * Example usage:
 * ```kotlin
 * import do_.mongo.Indexes.*
 *
 * // Single field ascending index
 * collection.createIndex(ascending("email"))
 *
 * // Compound index
 * collection.createIndex(compoundIndex(
 *     ascending("status"),
 *     descending("createdAt")
 * ))
 *
 * // Text index
 * collection.createIndex(text("content"))
 *
 * // Geospatial index
 * collection.createIndex(geo2dsphere("location"))
 * ```
 */
object Indexes {

    /**
     * Creates an ascending index on the specified fields.
     */
    fun ascending(vararg fields: String): Bson =
        Document(fields.associateWith { 1 })

    /**
     * Creates a descending index on the specified fields.
     */
    fun descending(vararg fields: String): Bson =
        Document(fields.associateWith { -1 })

    /**
     * Creates a 2dsphere geospatial index.
     */
    fun geo2dsphere(vararg fields: String): Bson =
        Document(fields.associateWith { "2dsphere" })

    /**
     * Creates a 2d geospatial index.
     */
    fun geo2d(field: String): Bson =
        Document(field, "2d")

    /**
     * Creates a text index on the specified fields.
     */
    fun text(vararg fields: String): Bson =
        Document(fields.associateWith { "text" })

    /**
     * Creates a text index on all string fields.
     */
    fun textAll(): Bson = Document("\$**", "text")

    /**
     * Creates a hashed index.
     */
    fun hashed(field: String): Bson =
        Document(field, "hashed")

    /**
     * Combines multiple index specifications into a compound index.
     */
    fun compoundIndex(vararg indexes: Bson): Bson {
        val combined = Document()
        for (index in indexes) {
            combined.putAll(index.toBsonDocument())
        }
        return combined
    }

    /**
     * Combines a collection of index specifications into a compound index.
     */
    fun compoundIndex(indexes: Collection<Bson>): Bson =
        compoundIndex(*indexes.toTypedArray())
}

/**
 * Index options builder.
 */
class IndexOptions {
    private val options = Document()

    /**
     * Sets the index name.
     */
    fun name(name: String): IndexOptions = apply {
        options["name"] = name
    }

    /**
     * Makes this a unique index.
     */
    fun unique(unique: Boolean = true): IndexOptions = apply {
        options["unique"] = unique
    }

    /**
     * Makes this a sparse index.
     */
    fun sparse(sparse: Boolean = true): IndexOptions = apply {
        options["sparse"] = sparse
    }

    /**
     * Makes this a background index (deprecated in MongoDB 4.2+).
     */
    fun background(background: Boolean = true): IndexOptions = apply {
        options["background"] = background
    }

    /**
     * Sets the TTL (time to live) in seconds.
     */
    fun expireAfter(seconds: Long): IndexOptions = apply {
        options["expireAfterSeconds"] = seconds
    }

    /**
     * Sets a partial filter expression.
     */
    fun partialFilterExpression(filter: Bson): IndexOptions = apply {
        options["partialFilterExpression"] = filter.toBsonDocument()
    }

    /**
     * Sets the collation.
     */
    fun collation(collation: Document): IndexOptions = apply {
        options["collation"] = collation
    }

    /**
     * Sets the default language for text indexes.
     */
    fun defaultLanguage(language: String): IndexOptions = apply {
        options["default_language"] = language
    }

    /**
     * Sets the language override field for text indexes.
     */
    fun languageOverride(field: String): IndexOptions = apply {
        options["language_override"] = field
    }

    /**
     * Sets text index weights.
     */
    fun weights(weights: Document): IndexOptions = apply {
        options["weights"] = weights
    }

    /**
     * Sets the 2dsphere index version.
     */
    fun sphereVersion(version: Int): IndexOptions = apply {
        options["2dsphereIndexVersion"] = version
    }

    /**
     * Sets the bits for 2d indexes.
     */
    fun bits(bits: Int): IndexOptions = apply {
        options["bits"] = bits
    }

    /**
     * Sets the min value for 2d indexes.
     */
    fun min(min: Double): IndexOptions = apply {
        options["min"] = min
    }

    /**
     * Sets the max value for 2d indexes.
     */
    fun max(max: Double): IndexOptions = apply {
        options["max"] = max
    }

    /**
     * Sets wildcard projection for wildcard indexes.
     */
    fun wildcardProjection(projection: Document): IndexOptions = apply {
        options["wildcardProjection"] = projection
    }

    /**
     * Makes this a hidden index.
     */
    fun hidden(hidden: Boolean = true): IndexOptions = apply {
        options["hidden"] = hidden
    }

    /**
     * Builds the options document.
     */
    fun build(): Document = Document(options)
}

/**
 * Creates index options using DSL syntax.
 */
inline fun indexOptions(block: IndexOptions.() -> Unit): Document =
    IndexOptions().apply(block).build()
